src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/AbstractLanguageServer.cs

```diff
@@ -116,29 +124,34 @@ protected virtual AbstractHandlerProvider HandlerProvider
     protected virtual void SetupRequestDispatcher(IHandlerProvider handlerProvider)
     {
         // Get unique set of methods from the handler provider for the default language.
-        foreach (var metadata in handlerProvider
+        foreach (var methodGroup in handlerProvider
             .GetRegisteredMethods()
-            .Select(m => new RequestHandlerMetadata(m.MethodName, m.RequestType, m.ResponseType, LanguageServerConstants.DefaultLanguageName))
-            .Distinct())
+            .GroupBy(m => m.MethodName))
         {
             // Instead of concretely defining methods for each LSP method, we instead dynamically construct the
             // generic method info from the exported handler types.  This allows us to define multiple handlers for
-            // the same method but different type parameters.  This is a key functionality to support TS external
-            // access as we do not want to couple our LSP protocol version dll to theirs.
-            //
-            // We also do not use the StreamJsonRpc support for JToken as the rpc method parameters because we want
-            // StreamJsonRpc to do the deserialization to handle streaming requests using IProgress<T>.
-
-            var method = DelegatingEntryPoint.GetMethodInstantiation(metadata.RequestType, metadata.ResponseType);
+            // the same method but different type parameters.  This is a key functionality to support LSP extensibility
+            // in cases like XAML, TS to allow them to use different LSP type definitions
+
+            // Verify that we are not mixing different numbers of request parameters and responses between different language handlers
+            // e.g. it is not allowed to have a method have both a parameterless and regular parameter handler.
+            var requestTypes = methodGroup.Select(m => m.RequestType);
+            var responseTypes = methodGroup.Select(m => m.ResponseType);
+            if (!(requestTypes.All(r => r is null) || requestTypes.Any(r => r is not null))
+                || !(responseTypes.All(r => r is null) || responseTypes.Any(r => r is not null)))
```

Comment:
Might need more coffee, but this logic is frying my brain. You're pretty confident in this? Because when I do the !(a || b) => !a && !b transformation to make it easier for me to understand, it doesn't seem to make sense.